lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -124,10 +130,72 @@ OpFoldResult ConstantOp::fold(ArrayRef<Attribute> operands) {
   return valueAttr();
 }
 
+/// Check if the results of \p op are of integer type and with known bitwidth.
+/// Can be used to determine if any fold is legal.
+bool hasKnownWidths(Operation *op) {
+  for (auto resTy : op->getResultTypes()) {
+    auto ty = resTy.cast<FIRRTLType>();
+    if (!ty.isa<IntType>() || ty.getBitWidthOrSentinel() == -1)
+      return false;
+  }
+  return true;
+}
+
+// Width extension of a constant int depends only on the sign of the operand.
+APInt extendConstantToWidth(APInt old, bool isSigned, unsigned newWidth) {
```

Comment:
Plz mark static, also should this be used by `constFoldFIRRTLBinaryOp` above instead of duplicating the logic?